File: 2022/day4/day4.js
const fs = require('fs');
const path = require('path');

function part1(input) {
  /** @type string[] */
  let lines = input.trim().split('\n');

  let overlap = 0;
  for (let line of lines) {
    let [a1, a2] = line.split(',');
    let s1 = a1.split('-').map(x => parseInt(x, 10));
    let s2 = a2.split('-').map(x => parseInt(x, 10));
    // s1 bigger than s2
    if (s1[0] <= s2[0] && s1[1] >= s2[1]) overlap++;
    // s1 bigger than s2
    else if (s2[0] <= s1[0] && s2[1] >= s1[1]) overlap++;
  }
  return overlap;
}

function part2(input) {
  /** @type string[] */
  let lines = input.trim().split('\n');

  let overlap = 0;
  for (let line of lines) {
    let [a1, a2] = line.split(',');
    let s1 = a1.split('-').map(x => parseInt(x, 10));
    let s2 = a2.split('-').map(x => parseInt(x, 10));
    let len1 = s1[1] - s1[0];
    let len2 = s2[1] - s2[0];

    // make sure s1 is smaller
    if (s1[0] > s2[0]) {
      [s1, s2] = [s2, s1];
      [len1, len2] = [len2, len1];
    }

    if (s1[0] + len1 >= s2[0]) overlap++;
  }
  return overlap;
}

let input = fs.readFileSync(path.resolve(__dirname, './input.txt'), 'utf8').replaceAll('\r', '');

// input = `
// 2-4,6-8
// 2-3,4-5
// 5-7,7-9
// 2-8,3-7
// 6-6,4-6
// 2-6,4-8`;

console.log('part1:', part1(input));
console.log('part2:', part2(input));

File: 2022/day4/day4.clj
(require '[clojure.string :as str]
         '[clojure.set :as set])

(def input-raw (slurp (str "2022/" (-> *file* (str/split #"/") butlast last) "/input.txt")))
#_(def input-raw "
2-4,6-8
2-3,4-5
5-7,7-9
2-8,3-7
6-6,4-6
2-6,4-8
")

(def input
  (->> input-raw
       str/trim
       str/split-lines
       (map (fn [line]
              (map (fn [num-range]
                     (->> (str/split num-range #"-") (map read-string)))
                   (str/split line #","))))))

(defn range-inclusive [start end] (range start (inc end)))

(defn part1 [input]
  (->> input
       (map (fn [pairs] (map #(apply range-inclusive %) pairs)))
       (filter (fn [[s1 s2]] 
                 (= (count (set/intersection (set s1) (set s2)))
                    (if (< (count s1) (count s2))
                      (count s1)
                      (count s2)))))
       count))

(defn part2 [input]
  (->> input
       (map (fn [pairs] (map #(apply range-inclusive %) pairs)))
       (filter (fn [[s1 s2]] (not= (count (concat s1 s2)) (count (distinct (concat s1 s2))))))
       count))

(prn "part1:" (part1 input))
(prn "part2:" (part2 input))
